src/target/target_flash.c

```diff
@@ -0,0 +1,271 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/* This file implements target flash interaction routines.
+ * Provides functionality for buffered flash operations
+ * It depends on target flash implementations
+ */
+
+#include "general.h"
+#include "target_internal.h"
+
+static int flash_buffered_write(target_flash_s *f, target_addr_t dest, const void *src, size_t len);
+static int flash_buffered_flush(target_flash_s *f);
+
+target_flash_s *target_flash_for_addr(target *t, uint32_t addr)
+{
+	for (target_flash_s *f = t->flash; f; f = f->next)
+		if ((f->start <= addr) && (addr < (f->start + f->length)))
+			return f;
+	return NULL;
+}
+
+static int target_enter_flash_mode(target *t)
+{
+	if (t->flash_mode)
+		return 0;
+
+	int ret = 0;
+	if (t->enter_flash_mode)
+		ret = t->enter_flash_mode(t);
+	else
+		/* Reset target on flash command */
+		/* This saves us if we're interrupted in IRQ context */
+		target_reset(t);
+
+	if (ret == 0)
+		t->flash_mode = true;
+
+	return ret;
+}
+
+static int target_exit_flash_mode(target *t)
+{
+	if (!t->flash_mode)
+		return 0;
+
+	int ret = 0;
+	if (t->exit_flash_mode)
+		ret = t->exit_flash_mode(t);
+	else
+		/* Reset target to known state when done flashing */
+		target_reset(t);
+
+	t->flash_mode = false;
+
+	return ret;
+}
+
+static int flash_prepare(target_flash_s *f)
+{
+	if (f->ready)
+		return 0;
+
+	int ret = 0;
+	if (f->prepare)
+		ret = f->prepare(f);
+
+	if (ret == 0)
+		f->ready = true;
+
+	return ret;
+}
+
+static int flash_done(target_flash_s *f)
+{
+	if (!f->ready)
+		return 0;
+
+	int ret = 0;
+	if (f->done)
+		ret = f->done(f);
+
+	if (f->buf) {
+		free(f->buf);
+		f->buf = NULL;
+	}
+
+	f->ready = false;
+
+	return ret;
+}
+
+int target_flash_erase(target *t, target_addr_t addr, size_t len)
+{
+	if (target_enter_flash_mode(t) != 0)
+		return 1;
+
+	int ret = 0;
+	while (len) {
+		target_flash_s *f = target_flash_for_addr(t, addr);
+		if (!f) {
+			DEBUG_WARN("Requested address is outside the valid range 0x%06" PRIx32 "\n", addr);
+			return 1;
+		}
+
+		/* terminate flash operations if we're not in the same target flash */
+		for (target_flash_s *target_f = t->flash; target_f; target_f = target_f->next)
+			if (target_f != f)
+				ret |= flash_done(target_f);
+
+		const target_addr_t local_start_addr = addr & ~(f->blocksize - 1U);
+		const target_addr_t local_end_addr = local_start_addr + f->blocksize;
+
+		if(flash_prepare(f) != 0)
+			return -1;
+
+		ret |= f->erase(f, local_start_addr, f->blocksize);
+
+		len -= MIN(local_end_addr - addr, len);
+		addr = local_end_addr;
+
+		/* Issue flash done on last operation */
+		if (len == 0)
+			ret |= flash_done(f);
+	}
+	return ret;
+}
+
+int target_flash_write(target *t, target_addr_t dest, const void *src, size_t len)
+{
+	if (target_enter_flash_mode(t) != 0)
+		return 1;
+
+	int ret = 0;
+	while (len) {
```

Comment:
Same query about using a for loop as above

Reply:
like above